src/dmd/root/strtold.d

```diff
@@ -0,0 +1,517 @@
+            goto Lerr;              // exponent is required
+    }
+    else
+    {
+        while (1)
+        {
+            int  i = *p;
+
+            while (isdigit(i))
+            {
+                ndigits = 1;        /* must have at least 1 digit   */
+                if (msdec < (0x7FFFFFFFFFFFL-10)/10)
+                    msdec = msdec * 10 + (i - '0');
+                else  if (msscale < (0xFFFFFFFF-10)/10)
+                {
+                    lsdec = lsdec * 10 + (i - '0');
+                    msscale *= 10;
+                }
+                else
+                {
+                    exp++;
+                }
+                exp -= dot;
+                i = *++p;
+            }
+            if (i == '.' && !dot)
+            {
+                p++;
+                dot++;
+            }
+            else
+                break;
+        }
+        if (!ndigits)               // if error (no digits seen)
+            goto Lerr;              // return 0.0
+    }
+    if (*p == 'e' || *p == 'E')
+    {
+        char  sexp;
+        int  e;
+
+        sexp = 0;
+        switch (*++p)
+        {
+        case '-':
+            sexp++;
+            goto case;
+        case '+':
+            p++;
+            break;
+        default:
+            break;
+        }
+        ndigits = 0;
+        e = 0;
+        while (isdigit(*p))
+        {
+            if (e < 0x7FFFFFFF / 10 - 10)   // prevent integer overflow
+            {
+                e = e * 10 + *p - '0';
+            }
+            p++;
+            ndigits = 1;
+        }
+        exp += sexp ? -e : e;
+        if (!ndigits)               // if no digits in exponent
+            goto Lerr;              // return 0.0
+    }
+
+    ldval = msdec;
+    if (msscale != 1)               /* if stuff was accumulated in lsdec */
+        ldval = ldval * msscale + lsdec;
+    if (ldval)
+    {
+        uint u = 0;
+        pow = 4096;
+
+        while (exp > 0)
+        {
+            while (exp >= pow)
+            {
+                ldval = ldval * postab[u];
+                exp -= pow;
+            }
+            pow >>= 1;
+            u++;
+        }
+        while (exp < 0)
+        {
+            while (exp <= -pow)
+            {
+                ldval = ldval * negtab[u];
+                if (ldval == 0)
+                    errno = ERANGE;
+                exp += pow;
+            }
+            pow >>= 1;
+            u++;
+        }
+        static if(0) {
+            for (int i = 0; i < 5; i++)
+                printf("%04x ",ldval.value[i]);
+            printf("\n");
+            printf("%llx\n",dval);
+        }
+    }
+L6: // if overflow occurred
+    if (ldval == longdouble_soft.infinity)
+        errno = ERANGE;
+
+L1:
+    if (endp)
+    {
+        *endp = cast(char *) p;
+    }
+L3:
+    return  sign ? -ldval : ldval;
+
+Lerr:
+    p = pinit;
+    goto L1;
+}
+
+/************************* Test ************************************/
+static if (0)
```

Comment:
remove ...

Reply:
I'm guessing the translated code should look as much as the original code as possible, to make it easier to review and find potential translation bugs. Then in follow up pull requests it can be enhanced and cleaned up.